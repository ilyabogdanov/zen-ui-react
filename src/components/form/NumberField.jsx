import React from "react";
import classNames from "classnames/bind";
import Utils from "../../Utils";

/**
 * @class
 * @classdesc Number Field
 * @author Ilya Bogdanov <public@ilyabogdanov.ru>
 * @license Apache-2.0
 *
 * @prop {boolean} props.disabled
 * @prop {boolean} props.error
 * @prop {boolean} props.stretch
 * @prop {number} props.value
 * @prop {string} props.placeholder
 * @prop {string} props.suffix
 *
 * @prop {*} props.children
 */
export default class NumberField extends React.Component {
    constructor(props) {
        // noinspection JSCheckFunctionSignatures
        super(props);
        this.state = {
            disabled: this.props.disabled,
            error: this.props.error,
            value: Number.isNaN(parseInt(this.props.value)) ? null : parseInt(this.props.value)
        };
        this.id = Utils.generateId();
        this._onClickPlus = this._onClick.bind(this, "+");
        this._onClickMinus = this._onClick.bind(this, "-");
        this._onChange = this._onChange.bind(this);
        this._onKeyPress = this._onKeyPress.bind(this);
    }
    _onKeyPress(e) {
        if (e.key==="+" || e.key==="-") {
            this._onClick(e.key);
        }
    }
    _onChange(e) {
        if (!this.state.disabled) {
            const value = Number.isNaN(parseInt(e.target.value)) ? null : parseInt(e.target.value);
            const valid = e.target.validity.valid;
            const onChange = this.props.onChange;
            if (valid && value !== this.state.value) {
                const result = value!==null ? parseInt(value) : null;
                this.setState({
                    value: result
                });
                onChange(result);
            }
        }
    }
    // noinspection JSCheckFunctionSignatures
    /**
     * @param {object} props
     * @param {boolean} props.error
     * @param {boolean} props.disabled
     */
    componentWillReceiveProps(props) {
        var newState = {};
        if (this.state.error!==props.error) {
            newState.error = props.error
        }
        if (this.state.disabled!==props.disabled) {
            newState.disabled = props.disabled
        }
        this.setState(newState);
    }
    _onClick(e) {
        if (!this.state.disabled) {
            const onChange = this.props.onChange;
            if (e==="+") {
                if (this.state.value===null) {
                    this.setState({
                        value: 1
                    });
                    onChange(1);
                } else {
                    const result = parseInt(this.state.value) + 1;
                    this.setState({
                        value: result
                    });
                    onChange(result);
                }
            } else if (this.state.value!==null && this.state.value!==0) {
                const result = parseInt(this.state.value) - 1;
                this.setState({
                    value: result
                });
                onChange(result);
            }
        }
    }
    render() {
        const { id, _onChange, _onKeyPress, _onClickPlus, _onClickMinus } = this;
        const { suffix, placeholder, stretch } = this.props;
        const { disabled, error, value } = this.state;
        return (
            <div className="zen_ui__number_field__container" data-stretch={stretch}>
                <div className="zen_ui__number_field" data-error={error}>
                    <div className="zen_ui__number_field__input_container">
                        <input pattern="[0-9]*"
                               onKeyPress={_onKeyPress}
                               onChange={_onChange}
                               disabled={disabled}
                               placeholder={placeholder}
                               id={id}
                               value={value || value===0 ? value : ""}/>
                        <div></div>
                    </div>
                    {
                        suffix ?
                            <div className="zen_ui__number_field__builtin_label">
                                <label htmlFor={id}>
                                    <div>{suffix}</div>
                                </label>
                            </div>
                        : null
                    }
                    <div className="zen_ui__number_field__buttons">
                        <div>
                            <div>
                                <div>
                                    <div>
                                        <label className={classNames("zen_ui__number_field__button__increase_number", {disabled: disabled})}
                                               htmlFor={id}
                                               onClick={_onClickPlus}></label>
                                    </div>
                                </div>
                                <div>
                                    <div>
                                        <label className={classNames("zen_ui__number_field__button__decrease_number", {disabled: disabled})}
                                               htmlFor={id}
                                               onClick={_onClickMinus}></label>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
};