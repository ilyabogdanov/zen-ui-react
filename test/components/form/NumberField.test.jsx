jest.dontMock("jquery");
import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import $ from "jquery";
import NumberField from "../../../src/components/form/NumberField";

class Container extends React.Component {
    constructor() {
        super();
        this.state = {
            disabled: false,
            error: false,
            value: 1
        };
        this._onChange = this._onChange.bind(this);
    }
    _onChange(e) {
        console.log(e);
    }
    render() {
        const { _onChange } = this;
        const { disabled, error, value } = this.state;
        return (
            <NumberField value={value} disabled={disabled} error={error} onChange={_onChange}/>
        )
    }
};

const div = document.createElement("div");
document.body.appendChild(div);

describe("Number Field", () => {

    // noinspection JSCheckFunctionSignatures
    const logSpy = jest.spyOn(console, "log");

    test("with parent should finish epic test", () => {

        const component = mount(
            <Container/>
        );
        expect(component.find(".zen_ui__number_field").prop("data-error")).toEqual(false);
        expect(component.find("input").prop("disabled")).toEqual(false);

        component.setState({ error: true });
        component.setState({ disabled: true });
        expect(component.find(".zen_ui__number_field").prop("data-error")).toEqual(true);
        expect(component.find("input").prop("disabled")).toEqual(true);

        component.setState({ disabled: false });
        component.find("input").simulate("change", {
            target: {
                value: 123,
                validity: {
                    valid: true
                }
            }
        });
        expect(component.find("input").prop("value")).toEqual(123);
        expect(logSpy).toHaveBeenCalledTimes(1);
        logSpy.mockReset();

        component.find("input").simulate("change", {
            target: {
                value: "xxx",
                validity: {
                    valid: true
                }
            }
        });
        expect(component.find("input").prop("value")).toEqual("");
        expect(logSpy).toHaveBeenCalledTimes(1);
        logSpy.mockReset();

        component.find("input").simulate("change", {
            target: {
                value: 666,
                validity: {
                    valid: false
                }
            }
        });
        expect(component.find("input").prop("value")).toEqual("");
        expect(logSpy).toHaveBeenCalledTimes(0);
        logSpy.mockReset();

        component.setState({ disabled: true });
        component.find("input").simulate("change", {
            target: {
                value: 456,
                validity: {
                    valid: true
                }
            }
        });
        expect(component.find("input").prop("value")).toEqual("");
        expect(logSpy).toHaveBeenCalledTimes(0);
        logSpy.mockReset();

        component.setState({ disabled: false });
        component.find("input").simulate("keyPress", {
            key: "+"
        });
        expect(component.find("input").prop("value")).toEqual(1);
        component.find("input").simulate("keyPress", {
            key: "+"
        });
        expect(component.find("input").prop("value")).toEqual(2);
        component.find("input").simulate("keyPress", {
            key: "-"
        });
        expect(component.find("input").prop("value")).toEqual(1);
        expect(logSpy).toHaveBeenCalledTimes(3);
        logSpy.mockReset();

        component.find("input").simulate("keyPress", {
            key: "k"
        });

    });
    test("should render correctly", () => {

        const e = renderer.create(
            <NumberField
                value={1}
                disabled={false}
                error={false}
                stretch={true}
                onChange={function () {}}/>
        ).toJSON();
        expect(e).toMatchSnapshot();

    });
    test("should change value with plus and minus buttons", () => {

        const wrapper = mount(
            <NumberField
                value={null}
                disabled={false}
                error={false}
                stretch={true}
                suffix="DD"
                placeholder="PLACEHOLDER"
                onChange={function () {}}/>,
            { attachTo: div }
        );
        expect(div.childNodes).toHaveLength(1);

        expect($("input")).toHaveLength(1);
        expect(wrapper.find("input").prop("value")).toEqual("");
        $(".zen_ui__number_field__button__increase_number").click();
        expect(wrapper.find("input").prop("value")).toEqual(1);
        $(".zen_ui__number_field__button__increase_number").click();
        expect(wrapper.find("input").prop("value")).toEqual(2);
        $(".zen_ui__number_field__button__decrease_number").click();
        expect(wrapper.find("input").prop("value")).toEqual(1);

        wrapper.setState({ disabled: true });
        $(".zen_ui__number_field__button__increase_number").click();
        expect(wrapper.find("input").prop("value")).toEqual(1);

        wrapper.setState({ disabled: false });
        $(".zen_ui__number_field__button__decrease_number").click();
        expect(wrapper.find("input").prop("value")).toEqual(0);
        $(".zen_ui__number_field__button__decrease_number").click();
        $(".zen_ui__number_field__button__increase_number").click().click();
        expect(wrapper.find("input").prop("value")).toEqual(2);

        wrapper.detach();
        expect(div.childNodes).toHaveLength(0);

    });
});