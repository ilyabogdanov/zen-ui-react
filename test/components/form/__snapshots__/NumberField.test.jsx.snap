// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Field should render correctly 1`] = `
<div
  className="zen_ui__number_field__container"
  data-stretch={true}
>
  <div
    className="zen_ui__number_field"
    data-error={false}
  >
    <div
      className="zen_ui__number_field__input_container"
    >
      <input
        disabled={false}
        id="zen-ui-id-2"
        onChange={[Function]}
        onKeyPress={[Function]}
        pattern="[0-9]*"
        placeholder={undefined}
        value={1}
      />
      <div />
    </div>
    <div
      className="zen_ui__number_field__buttons"
    >
      <div>
        <div>
          <div>
            <div>
              <label
                className="zen_ui__number_field__button__increase_number"
                htmlFor="zen-ui-id-2"
                onClick={[Function]}
              />
            </div>
          </div>
          <div>
            <div>
              <label
                className="zen_ui__number_field__button__decrease_number"
                htmlFor="zen-ui-id-2"
                onClick={[Function]}
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
